Add Apertus

EduardDurech opened the pull request:
Pre-release of Apertus from the Swiss AI Initiative

Main modifications from Llama
- xIELU Activation
- QK-norm

@ArthurZucker

haeggee committed:
init swissai model

haeggee committed:
AutoModelForCausalLM

haeggee committed:
AutoModelForCausalLM mapping

haeggee committed:
qk norm and post ln optional

haeggee committed:
fix wrong shape of qk norm: megatron uses head_dim

haeggee committed:
automodel fixes

haeggee committed:
minor fix in forward

EduardDurech committed:
`SwissAIForTokenClassification` support

dhia680 committed:
fix rope validation to accept llama3 scaling

EduardDurech committed:
Align `SwissAI` to v4.52.4

EduardDurech committed:
Align `SwissAI` to v4.53.1

EduardDurech committed:
Init CUDA xIELU

EduardDurech committed:
`SwissAI*`->`Apertus*`

EduardDurech commented:
@dhia680

EduardDurech committed:
ci fix

EduardDurech committed:
check_docstring ignore ApertusConfig

EduardDurech committed:
Licensing and placeholder tests

EduardDurech committed:
Placeholder doc

EduardDurech committed:
XIELU syntax

EduardDurech committed:
`_xielu_python` optimization

EduardDurech committed:
Fix xIELU

ArthurZucker reviewed (commented):
very nice and very transformers like! Do you mind using `modular` to isolate the changes?

EduardDurech commented:
> very nice and very transformers like! Do you mind using `modular` to isolate the changes?

Yep, I was planning to and Cyril suggested, I built from Alex's original implementation but I'll refactor

EduardDurech committed:
[tmp] `{beta,eps}` persistent=False
until {beta,eps} saved in checkpoint

EduardDurech committed:
Modular `Apertus`

EduardDurech committed:
CUDA xIELU logging

EduardDurech committed:
Merge upstream/main into model/apertus

EduardDurech committed:
ci fix

EduardDurech committed:
ci fix

EduardDurech committed:
ci fix

ArthurZucker commented:
do you want me to review this?

Cyrilvallez reviewed (commented):
Alright! Super clean implementation, very pleasing to review 🤗 Great work! I just added a few comments that should be super straightforward to implement 👌

Also, please rename all `past_key_value` to `past_key_values` (added `s` at the end) in modular, as we made sure the `s` is always present recently 👌
To make the consistency tests happy, please run `make fix-copies` at the root of your transformer fork! It will make sure to reapply modular correctly based on latest main!
I tried to do it myself quickly, but it appears your fork is in a protected org, so I couldn't push on it

Cyrilvallez reviewed (commented):
@ArthurZucker no worries, I'm reviewing this one! If you can just answer this question though, it would help 🤗

EduardDurech committed:
Update license

Co-authored-by: Cyril Vallez <cyril.vallez@gmail.com>

EduardDurech commented on src/transformers/models/apertus/modular_apertus.py:
```diff
@@ -0,0 +1,410 @@
+    pass
+
+
+class ApertusForTokenClassification(LlamaForTokenClassification):
+    pass
```
@Cyrilvallez some libraries we use require it

EduardDurech commented on tests/models/apertus/test_modeling_apertus.py:
```diff
@@ -0,0 +1,607 @@
+            pad=(0, max_cache_len - mask_1a.shape[-1]),
+            mode="constant",
+            value=torch.finfo(self.model_dtype).min,
+        )
+
+        _ = self.model.forward(
+            input_1a,
+            attention_mask=padded_mask_1a,
+            position_ids=position_ids_1a,
+            cache_position=torch.arange(part_a, device=torch_device),
+            past_key_values=past_key_values,
+        )
+
+        # forward run for the second part of input
+        input_1b = input_ids_shared_prefix[:, part_a:]
+        position_ids_1b = position_ids_shared_prefix[:, part_a:]
+        mask_1b = mask_shared_prefix[:, :, part_a:, :]
+
+        padded_mask_1b = torch.nn.functional.pad(
+            input=mask_1b, pad=(0, max_cache_len - mask_1b.shape[-1]), mode="constant", value=0
+        )
+
+        outs_1b = self.model.forward(
+            input_1b,
+            attention_mask=padded_mask_1b,
+            position_ids=position_ids_1b,
+            cache_position=torch.arange(
+                part_a,
+                input_ids_shared_prefix.shape[-1],
+                device=torch_device,
+            ),
+            past_key_values=past_key_values,
+        )
+        decoded_1b = [
+            self.tokenizer.decode(t)
+            for t in outs_1b.logits.argmax(-1)[
+                0, torch.where(position_ids_shared_prefix == position_ids_shared_prefix.max())[1] - part_a
+            ]
+        ]
+        self.assertEqual(decoded, decoded_1b)
```
Forwarding test refactoring to @dhia680 who will do them next week

dhia680 replied:
Done

EduardDurech committed:
Update tests/models/apertus/test_modeling_apertus.py

Co-authored-by: Cyril Vallez <cyril.vallez@gmail.com>

EduardDurech committed:
`.utils.import_utils.is_torchdynamo_compiling`

EduardDurech committed:
`Apertus` class ordering

EduardDurech committed:
`past_key_value{->s}`, `make fix-copies`

EduardDurech commented on src/transformers/activations.py:
```diff
@@ -185,6 +185,96 @@ def __getitem__(self, key):
+
+    def _xielu_cuda(self, x: Tensor) -> Tensor:
+        """Firewall function to prevent torch.compile from seeing .item() calls"""
+        original_shape = x.shape
+        # CUDA kernel expects 3D tensors, reshape if needed
+        while x.dim() < 3:
+            x = x.unsqueeze(0)
+        if x.dim() > 3:
+            x = x.view(-1, 1, x.size(-1))
+        if original_shape != x.shape:
+            logger.warning_once(
+                "Warning: xIELU input tensor expects 3 dimensions but got (shape: %s). Reshaping to (shape: %s).",
+                original_shape,
+                x.shape,
+            )
```
@Cyrilvallez wouldn't multimodal models? It's not currently implemented in the activation to handle but it's something we're adding

EduardDurech replied:
@Cyrilvallez I actually noticed this is necessary for SGLang, for some reason it sends 2D

EduardDurech committed:
ci fix

Cyrilvallez commented on src/transformers/models/apertus/modular_apertus.py:
```diff
@@ -0,0 +1,410 @@
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+        self.act_fn = ACT2FN[config.hidden_act]
+        if config.hidden_act != "xielu":
+            self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
```
Then let's remove the path entirely! We don't want to have code paths which will never be taken! If any user actually wants to change it, they can either use Llama directly, or fork the code!

Cyrilvallez commented on src/transformers/models/apertus/modular_apertus.py:
```diff
@@ -0,0 +1,410 @@
+class ApertusAttention(LlamaAttention):
+    def __init__(self, config: ApertusConfig, layer_idx: Optional[int] = None):
+        super().__init__(config, layer_idx)
+        if self.config.qk_norm:
+            self.q_norm = ApertusRMSNorm(self.head_dim, config.rms_norm_eps)
+            self.k_norm = ApertusRMSNorm(self.head_dim, config.rms_norm_eps)
+        else:
+            self.q_norm = nn.Identity()
+            self.k_norm = nn.Identity()
```
See my answer above - let's remove the path!

EduardDurech committed:
Remove unused configuration parameters

EduardDurech committed:
`{beta,eps}` saved in checkpoint

chiffa commented:
@ArthurZucker @Cyrilvallez  - terribly sorry to bother you again, but is there anything we have to do on our side to finalize this PR? We really would need to start testing internally soon and start propagating to other repositories (eg vLLM, ...)

Cyrilvallez reviewed (commented):
Hey! It's all ready! Super nice work! Let's simply implement those small final comments, fix the conflict, then I'll merge! 🤗🤗

Cyrilvallez commented:
Also, please add the model to `docs/source/en/_toctree.yml` for our doc builder 🤗

Cyrilvallez commented on tests/models/apertus/test_modeling_apertus.py:
```diff
@@ -0,0 +1,344 @@
+    )
+    test_headmasking = False
+    test_pruning = False
+    fx_compatible = False  # Broken by attention refactor cc @Cyrilvallez
```
This can be removed, it's False by default and is now deprecated

dhia680 replied:
Done

Cyrilvallez commented on tests/models/apertus/test_modeling_apertus.py:
```diff
@@ -0,0 +1,344 @@
+    test_pruning = False
+    fx_compatible = False  # Broken by attention refactor cc @Cyrilvallez
+    model_tester_class = ApertusModelTester
+    rotary_embedding_layer = ApertusRotaryEmbedding  # Enables RoPE tests if set
```
This should be removed as well, as the tests are written for a config without default scaling (which is not the case in Apertus) - besides, repo scaling is tested in other models so it's all safe

EduardDurech committed:
`{beta,eps}` Temporarily on CPU

EduardDurech committed:
Suggestions

Co-authored-by: Cyril Vallez <cyril.vallez@gmail.com>

EduardDurech committed:
Merge branch 'main' into model/apertus

EduardDurech committed:
ci fix

dhia680 committed:
remove fx_compatible (deprecated)

dhia680 committed:
remove `rotary_embedding_layer`

As the tests are written for a config without default scaling (which is not the case in Apertus) - besides, rope scaling is tested in other models so it's all safe.

dhia680 committed:
fully removing `Mask4DTestHard` class

Not needed (for now)

dhia680 committed:
switch to `dtype` instead of `torch_dtype`

Following this:
https://github.com/huggingface/transformers/pull/39782

dhia680 commented on docs/source/en/model_doc/apertus.md:
```diff
@@ -0,0 +1,100 @@
+pipeline = pipeline(
+    task="text-generation",
+    model="swiss-ai/Apertus-8B",
+    torch_dtype=torch.bfloat16,
```
@Cyrilvallez I saw your [PR](https://github.com/huggingface/transformers/pull/39782) and updated the argument accordingly in the test script.
However, I still see `torch_dtype` used in some places such as when clicking on `use this model` from the model card on HF and choosing `transformers` as a library.
Will `torch_dtype` become deprecated ?

Cyrilvallez replied:
You can consider it as deprecated already - however, its usage will keep being valid for some time as it's a very popular argument - we need to ensure a very smooth transition

dhia680 committed:
remove unused imports

dhia680 commented on docs/source/en/model_doc/apertus.md:
```diff
@@ -0,0 +1,100 @@
+)
+input_ids = tokenizer("Plants create energy through a process known as", return_tensors="pt").to("cuda")
+
+output = model.generate(**input_ids, cache_implementation="static")
```
makes sense

dhia680 committed:
remove `cache_implementation="static"`

dhia680 committed:
+Apertus to `docs/source/en/_toctree.yml` for the doc builder

Cyrilvallez reviewed (approved):
Alright! Very very nice, congrats! Super efficient work! Merging it now! 🤗
